middle_end/flambda2/types/provers.ml

```diff
@@ -495,7 +495,10 @@ let meet_is_flat_float_array env t : bool meet_shortcut =
   | Value Unknown -> Need_meet
   | Value Bottom -> Invalid
   | Value (Ok (Array { element_kind = Unknown; _ })) -> Need_meet
-  | Value (Ok (Array { element_kind = Known element_kind; _ })) -> (
+  | Value (Ok (Array { element_kind = Bottom; _ })) ->
+    (* Empty array case. We cannot return Bottom, but any result is correct. *)
```

Comment:
True, but returning `false` here seems more consistent with the fact that the empty array always has tag zero, right?

Reply:
Actually I'm a bit worried about the code in `Simplify_common.specialise_array_kind`. I may have to return `Need_meet` just to make sure that we're not turning valid array operations on an empty array into `Invalid`.

Reply:
I've expanded the comment here a bit, and added comments at the use points too.